src/linker/Linker/CompilerGeneratedState.cs

```diff
@@ -165,6 +185,77 @@ void ProcessMethod (MethodDefinition method)
 					}
 				}
 			}
+
+			// Now that we have instantiating methods fully filled out, walk the state machines and fill in the attribute
+			// providers
+			foreach (var stateMachineType in _generatedTypeToTypeArgumentInfo.Keys) {
+				if (!stateMachineType.HasGenericParameters) {
+					continue;
+				}
+				MapGeneratedTypeTypeParameters (stateMachineType);
+			}
+
+			return;
+
+			void MapGeneratedTypeTypeParameters (TypeDefinition stateMachineType)
+			{
+				var typeInfo = _generatedTypeToTypeArgumentInfo[stateMachineType];
+				if (typeInfo.OriginalAttributes is not null) {
+					return;
+				}
+				var method = typeInfo.CreatingMethod;
+				if (method.Body is { } body) {
+					var typeArgs = new ICustomAttributeProvider[stateMachineType.GenericParameters.Count];
```

Comment:
Nit: Move this after the ScanForInit and if - since it only needs to be allocated when we're actually going to use it.